src/Axe/Controllers/CompilerController.cs

```diff
@@ -46,80 +51,59 @@ public IActionResult Solve(int id)
         }
 
         [HttpPost]
-        public IActionResult Solve(CodeBlock codeBlock)
+        public IActionResult Solve(CodeBlockVm model)
         {
+            System.Threading.Thread.Sleep(2000);
+            model.SourceCode = compileManager.FormatCode(model.SourceCode);
             if (ModelState.IsValid)
             {
                 try
                 {
-                    string assemblyName = "AxeCompilerAssembly" + Path.GetRandomFileName();
-                    // Random file name is necessary because .NETCore doesn't allow to unload assembly
-                    SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(codeBlock.SourceCode);
-
-                    StatementSyntax syntaxOfVerificationCode = SyntaxFactory.ParseStatement(codeBlock.VerificationCode.Replace("#number#", "AXE"));
-                    MethodDeclarationSyntax methodDeclaration = SyntaxFactory.MethodDeclaration(SyntaxFactory.ParseTypeName("bool"), "CHECK")
-                        .AddModifiers(SyntaxFactory.Token(SyntaxKind.PublicKeyword))
-                        .WithBody(SyntaxFactory.Block(syntaxOfVerificationCode));
-
-                    ClassDeclarationSyntax axeTaskClass = syntaxTree.GetRoot().DescendantNodes()
-                        .OfType<ClassDeclarationSyntax>()
-                        .First(c => c.Identifier.ValueText == "AxeTask");
-
-                    ClassDeclarationSyntax newAxeTaskClass = axeTaskClass.AddMembers(methodDeclaration);
-                    SyntaxNode syntaxNode = syntaxTree.GetRoot().ReplaceNode(axeTaskClass, newAxeTaskClass);
-                    string sourceCode = syntaxNode.NormalizeWhitespace().ToFullString();
-                    syntaxTree = CSharpSyntaxTree.ParseText(sourceCode);
-
-                    MetadataReference[] references = new MetadataReference[]
-                    {
-                        MetadataReference.CreateFromFile(typeof(object).GetTypeInfo().Assembly.Location),
-                        MetadataReference.CreateFromFile(typeof(Hashtable).GetTypeInfo().Assembly.Location)
-                    };
-
-                    CSharpCompilation compilation = CSharpCompilation.Create(
-                        assemblyName: assemblyName,
-                        syntaxTrees: new[] { syntaxTree },
-                        references: references,
-                        options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
-
-                    using (var stream = new MemoryStream())
+                    Tuple<CodeBlockResult, string[]> result = compileManager.Solve(model);
+                    model.Result = result.Item1;
+                    if (result.Item1 == CodeBlockResult.Error || result.Item1 == CodeBlockResult.Failed)
```

Comment:
I'm still expecting when this post-processing is moved to manager before accepting pull-request

Reply:
I already all implementation is moved to CompileManager. The Solve method returns Tuple where the first item is result, the second is string array of errors. Thus, you can use unit tests to test this method. In controllers I do the only action - to prepare model for view (add errors to ModelState if they are)